tensorflow/python/ops/math_ops.py

```diff
@@ -3956,6 +3956,24 @@ def add(x, y, name=None):
   with ops.name_scope(name, "Add", [x]) as name:
     x = ops.convert_to_tensor(x, name="x")
     y = ops.convert_to_tensor(y, dtype_hint=x.dtype.base_dtype, name="y")
+    
+    # Bug fix: Prevent overflow/underflow when mixing integer types
+    if (x.dtype.is_integer and y.dtype.is_integer and x.dtype != y.dtype):
+      # Determine safe dtype based on combined value ranges
+      combined_range = max(abs(x.dtype.min), abs(x.dtype.max), 
```

Comment:
If `x` and `y` are int8 and int16, the max value will be `abs(-32768)` which is +32768, and be converted below to int32.  Is this intentional?  This may change behavior of existing workflows.